File: Javascript/Binary Search Tree/bst_traversal.js
function inOrderTraverse(tree, array) {
	if (!tree) return array;

	inOrderTraverse(tree.left, array);
	array.push(tree.value);
	inOrderTraverse(tree.right, array);

	return array;
}

function preOrderTraverse(tree, array) {
	if (!tree) return array;

	array.push(tree.value);
	preOrderTraverse(tree.left, array);
	preOrderTraverse(tree.right, array);

	return array;
}

function postOrderTraverse(tree, array) {
	if (!tree) return array;

	postOrderTraverse(tree.left, array);
	postOrderTraverse(tree.right, array);
	array.push(tree.value);

	return array;
}

// Time Complexity: O(n)
// Space Complexity: O(n), if no array, the space complexity would be O(d
// where d is the depth of the BST

File: AE/Binary Search Trees/bst_traversal.py
def in_order_traversal(tree, array):
    if tree is None:
        return

    in_order_traversal(tree.left, array)
    array.append(tree.value)
    in_order_traversal(tree.right, array)

    return array


def pre_order_traversal(tree, array):
    if tree is None:
        return

    array.append(tree.value)
    pre_order_traversal(tree.left, array)
    pre_order_traversal(tree.right, array)

    return array


def post_order_traversal(tree, array):
    if tree is None:
        return

    post_order_traversal(tree.left, array)
    post_order_traversal(tree.right, array)
    array.append(tree.value)

    return array

# Time Complexity: O(n)
# Space Complexity: O(n)
